### ISSUE TITLE
Add tasks without unlocking phone

### ISSUE THREAD

**Issue by djibux** [mpcjanssen/simpletask-android#583]
On the bottom of the lockscreen I can add a button called "Add Task" that takes me straight to Simpletask, but asks for my unlock pin code.

I wish I could add tasks without unlocking the phone just like I can take pictures without unlocking.

I am using CyanogenMod 13 (Android 6.0.1).

**Comment by smichel17** [comment]
You can! You just need a way to launch the AddTask screen without unlocking your phone. Eg, I [forked AcDisplay](https://gitlab.com/smichel17/AcDisplay) and modified it to launch Simpletask/Addtask and Calendar/Add Appointment instead of phone and camera, upon corner swipe.

**Comment by djibux** [comment]
So it needs a custom app?  This cannot be native?

**Comment by smichel17** [comment]
The problem isn't running Simpletask from the lock screen (that already exists), it's finding a way to launch it from the lock screen.

If you can find any way of getting a launcher on to your lock screen (like I did with AcDisplay), it'll work just fine.

**Comment by djibux** [comment]
Well I can't seem to be able to run Simpletask from the lock screen.

As I wrote I can customize my lockscreen and thus add the AddTask button easily.  But when I then click on that button, the phone asks for my pin.

**Comment by smichel17** [comment]
Oh, I missed that, my apologies. I'll take a look at this again when I'm at my computer.

**Comment by djibux** [comment]
No worries ;) Thanks.

Le 10 septembre 2016 23:02:00 GMT+02:00, smichel17 <notifications@github.com> a écrit :
>Oh, I missed that, my apologies. I'll take a look at this again when
>I'm at my computer. 
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/mpcjanssen/simpletask-android/issues/583#issuecomment-246140146

-- 
djib

**Comment by smichel17** [comment]
@djibux Is the acdisplay fork working for you?

Also out of curiosity, are you able to run other apps from lock screen the way you were trying to run simpletask?

**Comment by djibux** [comment]
@smichel17 could you provide an apk for the acdisplay fork?  I can't seem to be able to compile it (it says `Could not find com.android.tools.build:gradle:2.2.0-alpha3`.

The camera is the only application I know of that I can start without unlocking the phone.  AcDisplay (original version) also works with the screen locked though it is a bit different: I doesn't require to be launched…

**Comment by smichel17** [comment]
You need to `git reset --hard` back to the commit I made.

**Comment by smichel17** [comment]
There is a debug build up here though: http://smichel.me/acdisplay-smichel17.apk

**Comment by djibux** [comment]
Thanks.  I managed to build by reseting to fa00cfc70a5b7a066d0db4a3f6adcd61854d106a.  Guess what: it works!  I am a bit confused as to why it doesn't work from the native CyanogenMod lock screen…

**Comment by smichel17** [comment]
On Android, each activity has a couple flags that can be set to allow it to run when the screen is locked. Simpletask's AddTask activity has those flags; it's main activity does not (and neither do the filter modification, lua config, or preference activities).

Launching from lockscreen is done via intent. There are a couple different kinds of intent. The one I used launches the AddTask activity directly. I don't know the details of all the kinds of intents or exactly how simpletask handles them; my best guess at why the lockscreen doesn't work is that the widget might open the AddTask activity indirectly, through an activity that isn't flagged to run on the lock screen.

**Comment by mpcjanssen** [comment]
@smichel17 that's probably what's happening.

**Comment by smichel17** [comment]
@djibux there's an app called 'activity launcher' on fdroid that will allow you to create a shortcut directly to the activity. Not sure if you'll be able to get it on the lock screen but worth a shot if you don't want to use acdisplay. I use it to create a one-click `new calendar appointment` button when I'm not on the lock screen and it is very nice :)

**Comment by djibux** [comment]
Thanks @smichel17. Unfortunately I can't seem to be able to add "activity launcher" shortcuts to the home screen.

I am not quite sure how I can held debug that issue.  It might actually be a CyanogenMod issue because I am selecting the "Add Task" activity as a shortcut on the lockscreen, I am not using a widget.

**Comment by ddisanti** [comment]
@djibux mayybe you can try  Xposed Framework and these modules:

- [AOSP][LP-MM] Gesture Unlock with Custom Shortcuts
http://repo.xposed.info/module/me.rijul.gestureunlock

- Shortcut Lock Screen Mod [MM]
http://repo.xposed.info/module/com.android.xposed.shortcutlockscreenmod

I have tried the first one and it works fine, but gestures are not for me. :smile:

**Comment by smichel17** [comment]
@djibux If you're going to use Xposed, "App Settings" which I mentioned/linked earlier also includes the ability to allow arbitrary apps to run on the lock screen.

**Comment by smichel17** [comment]
Finally, we can also investigate how widgets launch the AddTask activity and make it so that process can happen while the device is locked.

**Comment by djibux** [comment]
Thanks for the tip @ddisanti. It seems Xposed Framework require a rooted device, which I'm not too keen on doing.  But as I understand from your message you have managed to run AddTask from you lockscreen using the first module so it might indeed be an issue with CyanogenMod rather than simpletask.

@smichel17 Is there anything I can do like sharing a logcat or something?

**Comment by smichel17** [comment]
@djibux I don't think so. It just requires time to look throng the code and see how that widget works.

**Comment by ddisanti** [comment]
@djibux ok. CyanogenMod isn't rooted already?  At least it was in my old Samsung S2 days. Oh. Your device came with CyanogenMod installed.

**Comment by djibux** [comment]
@ddisanti well, in the current version you can select either: no root access, adb root access, application root access or full root access.  I have selected the first option.

**Comment by smichel17** [comment]
Looked into this. It's a "bug" with CM lockscreen, not simpletask. I don't know quite enough yet; it *might* be possible to create a workaround, or this might be a wontfix. Here's the deal:

Android has a built-in feature for creating shortcut widgets. You send an intent X to the android system that basically says, "I want to create a shortcut with icon Y and when the user clicks it, it launches intent Z", and then the android system creates the shortcut for you. We have no control over this working on the lock screen because simpletask isn't creating the widget, the android system (or maybe the CM lockscreen; I don't know how their implementation works) is.

It might be possible to create the widget directly ourselves (like how we create the list widget), in which case we can make it work on the lock screen. The thing is, I am not sure if that widget would be as flexible as the current one. For example, you can (and I did) put the current addtask widget in the bottom row of your home screen, that follows you from page to page, but you can't put the list widget there.

@djibux Can you put the list widget on your lockscreen? If so, does creating a new task with the `+` in its upper right corner work without unlocking your phone?

**Comment by djibux** [comment]
Thanks for looking into this @smichel17.  As far as I know I cannot add widgets to my lock screen, only customize the two bottom corner actions (phone and camera by default).  I have changed the phone button to put the AddTask activity.

**Comment by smichel17** [comment]
Oh, I see  misunderstood what you were talking about; I'm not familiar with CM because my phone is unfortunately locked down with respect to custom roms (galaxy s4 through verizon; won't be doing that for my next phone).

If CM lockscreen is just launching the AddTask activity it seems like this should work... if that's what you think it is doing but it still requires unlock to launch, a logcat would be helpful to see what's actually going on.

**Comment by djibux** [comment]
Here is a pastebin.  https://framabin.org/?4517ad908667133b#SG8SCqvGGrizjZBgnDz9H9pOUh8j6m43UPfrpdql7+A=

I have put a comment in place where I launched the activity by sliding from the bottom left.  The next logs are about unlocking the phone before getting to the line
```
   10-02 15:44:34.764 3648 5365 I ActivityManager: START u0 {act=android.intent.action.MAIN flg=0x14000000 cmp=nl.mpcjanssen.simpletask/.AddTask (has extras)} from uid 10018 on display 0

**Comment by smichel17** [comment]
This is beyond my ability to fix within a reasonable effort-benefit ratio.

**Comment by djibux** [comment]
OK, I will report a bug to CyanogenMod.

Le 17 octobre 2016 03:13:57 GMT+02:00, smichel17 <notifications@github.com> a écrit :
>This is beyond my ability to fix within a reasonable effort-benefit
>ratio.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/mpcjanssen/simpletask-android/issues/583#issuecomment-254091417

-- 
djib

**Comment by smichel17** [comment]
I have a samsung galaxy S6 running Android 6.0 that allows customizing the bottom corner actions, but it only allows starting the main activity of the app, not the AddTask activity.
